java/ql/lib/semmle/code/java/frameworks/android/Android.qll

```diff
@@ -5,6 +5,25 @@
 import java
 private import semmle.code.xml.AndroidManifest
 
+/**
+ * There is an android manifest file which defines an activity, service or
+ * content provider (so it corresponds to an android application rather than a
+ * library), and `file` is in a subfolder of the folder that contains it.
+ */
+predicate inAndroidApplication(File file) {
+  file.isSourceFile() and
+  exists(AndroidComponentXmlElement acxe, AndroidManifestXmlFile amxf |
+    amxf.getManifestElement().getApplicationElement().getAComponentElement() = acxe and
+    (
+      acxe instanceof AndroidActivityXmlElement or
+      acxe instanceof AndroidServiceXmlElement or
+      acxe instanceof AndroidProviderXmlElement
+    )
+  |
+    file.getParentContainer+() = amxf.getParentContainer()
+  )
+}
+
```

Comment:
```suggestion
/**
 * Holds if `amxf` defines at least one activity, service or contest provider,
 * and so it corresponds to an android application rather than a library.
 */
private predicate definesAndroidApplication(AndroidManifestXmlFile amxf) {
  exists(AndroidComponentXmlElement acxe |
    amxf.getManifestElement().getApplicationElement().getAComponentElement() = acxe and
    (
      acxe instanceof AndroidActivityXmlElement or
      acxe instanceof AndroidServiceXmlElement or
      acxe instanceof AndroidProviderXmlElement
    )
  )
}

/**
 * Holds if in `file`'s directory or some parent directory there is an `AndroidManifestXmlFile`
 * that defines at least one activity, service or contest provider, suggesting this file is
 * part of an android application.
 */
predicate inAndroidApplication(File file) {
  file.isSourceFile() and
  exists(AndroidManifestXmlFile amxf, Folder amxfDir |
    definesAndroidApplication(amxf) and amxfDir = amxf.getParentContainer()
  |
    file.getParentContainer+() = amxfDir
  )
}

```

Reply:
(rewords the comment to use `Holds if`, and splits it into two predicates)